qemu-vm: remove bootDisk, refactor using make-disk-image with better EFI support

RaitoBezarius opened the pull request:
### Description of changes

This PR does multiple things :

###### make-disk-image refactoring in QEMU test infrastructure

As per https://github.com/NixOS/nixpkgs/issues/23052 — QEMU test infrastructure do not use ad-hoc code to build the boot disk image but reuse make-disk-image which was improved to support use-cases.

Doing this consolidates make-disk-image as one place to do these changes, it seems like the boot disk production code was doing old stuff with GRUB 0.97, I'm uncertain whether that should be cleaned up.

###### UEFI firmware infrastructure

As part of https://github.com/NixOS/nixpkgs/pull/187887 ; this uses consolidated UEFI firmware and variables and expose their variables to end users.

It also adds SecureBoot and System Management Mode enforcement support in the test infrastructure and the make-disk-image function.

EFIVARS can be modified as part of the disk image build process, enabling people to build a disk image with specific authenticated UEFI variables, e.g. SecureBoot keys.

###### Documentation

This PR adds documentation on what are the breaking changes in regard to the disk layout (storeImage is used only if needed at all.)

TODO/notes:

- [x] Cleanup UEFI firmware infrastructure: done in
- [x] Cleanup the /dev/vdb2 stuff in QEMU code.
- [x] Some tests such as installer depends on bootDevice being vdb or second drive.
- [x] GRUB version 1 may have some subtleties.
- [x] Assert touchEFIVars only if partition table type is hybrid or efi.
- [x] Use a better error message for assert.
- [x] Refactor in make-disk-image the EFI firmware infrastructure using https://github.com/NixOS/nixpkgs/pull/187887.
- [x] Simple tests:
- [x] Legacy boot without any EFI stuff should still work.
- [x] EFI/Hybrid boot without touching EFI vars should still work.
- [x] EFI/Hybrid boot with touching EFI vars should still work AND we should be able to write/read in EFIVARS.
- [x] Reproducibility of system image, wrt to fixed GUIDs in partition table.
- [x] Documentation:
    - [x] How to test this kind of changes? What to look for?
    - [x] Where are the breaking changes? Everything that relies on /dev/vdb2 is now incorrect.
    - [x] Compatibility layer? Deprecation notice?
    - [x] Prevent users to shoot their foot: increase the number of assertions & warnings accordingly in the module
- [ ] For a future PR
    - [ ] Over {systemd-boot, GRUB1, GRUB2} cartesian product for SecureBoot test.
    - [ ] makeEc2Test is useful to initialize using an image, it would be good to extract it as a lower level primitive to start test off an already existing disk image, maybe put it directly into makeTest machinery or makeTestFromImage

RaitoBezarius commented:
cc @baloo @roberth (the old one #187639 mass-pinged all the contributors due to a fail in the rebase.)

RaitoBezarius commented:
@ofborg test installer.simple
@ofborg test simple
@ofborg test non-default-filesystems
@ofborg test swap-partition
@ofborg test systemd-boot.basic
@ofborg test systemd-boot.update
@ofborg test systemd-boot.switch-test
@ofborg test secureboot.basic

roberth commented:
https://hydra.nixos.org/eval/1781068?compare=1780016 compares [94b75a2](https://github.com/NixOS/nixpkgs/commit/94b75a230b29810badaaa1fcb745a33d175740de) against the latest common commit.

fricklerhandwerk commented:
@RaitoBezarius please re-add me to reviewers once there are changes to documentation.

RaitoBezarius commented:
@ofborg test installer.simple
@ofborg test simple
@ofborg test non-default-filesystems
@ofborg test swap-partition
@ofborg test systemd-boot.basic
@ofborg test systemd-boot.update
@ofborg test systemd-boot.switch-test
@ofborg test secureboot.basic

RaitoBezarius commented:
@danielfullmer I don't know if you have noticed this, but it seems like latest systemd-boot has changed his version format, I thought it was related to my PR, but it was more due to a systemd-boot bump, see https://github.com/NixOS/nixpkgs/pull/191665/commits/61e53905bc19fc11b07c86bed115b3a7fe882f7e — I suppose it would be nice to backport this to master, but I was not able to `nix-build systemd-boot.nix -A update` on master, @roberth do you have an idea if there were changes in how NixOS tests are supposed to be run?

RaitoBezarius commented:
@ofborg test installer.simple
@ofborg test simple
@ofborg test non-default-filesystems
@ofborg test swap-partition
@ofborg test systemd-boot.basic
@ofborg test systemd-boot.update
@ofborg test systemd-boot.switch-test
@ofborg test secureboot.prevent-reboot
@ofborg test lvm2
@ofborg test hibernate
@ofborg test systemd-initrd-btrfs-raid
@ofborg test systemd-initrd-luks-keyfile
@ofborg test systemd-initrd-luks-password
@ofborg test systemd-initrd-sw-raid

roberth commented:
> but I was not able to `nix-build systemd-boot.nix -A update` on master, @roberth do you have an idea if there were changes in how NixOS tests are supposed to be run?

Yes, I've broken and fixed this way of invoking for files that import `make-test-python.nix`, but not yet for `testing-python.nix`. I'll restore the ugly workaround for "`nix-build <test.nix>`" that I've cleaned up.

`nix-build -A nixosTests.*` and hydra have continued to work as usual.

RaitoBezarius commented:
> > but I was not able to `nix-build systemd-boot.nix -A update` on master, @roberth do you have an idea if there were changes in how NixOS tests are supposed to be run?
> 
> Yes, I've broken and fixed this way of invoking for files that import `make-test-python.nix`, but not yet for `testing-python.nix`. I'll restore the ugly workaround for "`nix-build <test.nix>`" that I've cleaned up.
> 
> `nix-build -A nixosTests.*`, hydra and ofborg have continued to work as usual.

Alright! Can I ask you to restart the tests on Hydra for this PR? I cleaned up a lot of stuff and I think we are more or less good excluding already failing tests.

RaitoBezarius commented:
@ofborg test installer.simple
@ofborg test simple
@ofborg test non-default-filesystems
@ofborg test swap-partition
@ofborg test systemd-boot.basic
@ofborg test systemd-boot.update
@ofborg test systemd-boot.switch-test
@ofborg test secureboot.prevent-reboot
@ofborg test lvm2
@ofborg test hibernate
@ofborg test systemd-initrd-btrfs-raid
@ofborg test systemd-initrd-luks-keyfile
@ofborg test systemd-initrd-luks-password
@ofborg test systemd-initrd-swraid

RaitoBezarius commented:
@fricklerhandwerk I added documentation, but I'm not certain about their relevance in some places, see `make-disk-image` documentation, I can add more code documentation if you feel that's interesting, I am planning to do a "How to contribute to QEMU test architecture" document in a later PR to help with this kind of changes.

RaitoBezarius commented:
@ofborg test installer.simple
@ofborg test simple
@ofborg test non-default-filesystems
@ofborg test swap-partition
@ofborg test systemd-boot.basic
@ofborg test systemd-boot.update
@ofborg test systemd-boot.switch-test
@ofborg test secureboot.prevent-reboot
@ofborg test lvm2
@ofborg test hibernate
@ofborg test systemd-initrd-btrfs-raid
@ofborg test systemd-initrd-luks-keyfile
@ofborg test systemd-initrd-luks-password
@ofborg test systemd-initrd-swraid

RaitoBezarius commented:
For curious people, here's a determinism analysis on the root partition in a `legacy+gpt` disk image, you can see that GUIDs are fixed: https://files.lahfa.xyz/private/diffoscope-make-disk-image-ext4-diff.html (as GitHub do not enable HTML uploading.)

fricklerhandwerk commented:
Thanks for your work, it looks great from a distance, but going into detail is too much for me to handle right now, as I'm focused on other topics. Either you'd wait 1-2 weeks (although I can't guarantee anything) or ping someone else and I'll unsubscribe.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1287

roberth commented:
Baseline eval e6d6133ee8ec2c5e26398cbfb46e2c4f7ee42565: https://hydra.nixos.org/eval/1783709
PR eval 7383fca9d36d9af1dcf104f1f77c9a5f47a7afbf: in progress... https://hydra.nixos.org/jobset/nixos/python-test-refactoring

SuperSandro2000 commented on nixos/tests/systemd-initrd-luks-password.nix:
```diff
@@ -43,6 +43,6 @@ import ./make-test-python.nix ({ lib, pkgs, ... }: {
-    assert "/dev/mapper/cryptroot on / type ext4" in machine.succeed("mount")
-  '';
+    assert "/dev/mapper/cryptroot on / type ext4" in machine.succeed("mount"), "/dev/mapper/cryptroot do not appear in mountpoints list"
+    '';
```
```suggestion
  '';
```

SuperSandro2000 commented on nixos/tests/secureboot.nix:
```diff
@@ -0,0 +1,42 @@
+    };
+    boot.loader.systemd-boot.enable = true;
+    boot.loader.efi.canTouchEfiVariables = true;
+    environment.systemPackages = [ pkgs.efibootmgr pkgs.sbsigntool pkgs.sbctl ];
```
```suggestion
    environment.systemPackages = with pkgs; [ efibootmgr sbsigntool sbctl ];
```

SuperSandro2000 commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -360,10 +307,35 @@ in
+          then "${cfg.bootDevice}2"
+          else
+            if !cfg.useBootLoader then cfg.bootDevice else null;
+        defaultText = ''if (cfg.useBootLoader && !hasPrefix "/dev/mapper/" ''${cfg.bootDevice}) then "''${cfg.bootDevice}2" else cfg.bootDevice;'';
```
Maybe I am missing something but that is not matching default 1:1.

SuperSandro2000 commented on nixos/lib/make-disk-image.nix:
```diff
@@ -405,7 +477,17 @@ let format' = format; in let
+        ${optionalString (config.boot.loader.grub.device != "/dev/vda") ''
+            ln -s /dev/vda ${config.boot.loader.grub.device}
+        ''}
+        # TODO: needed?
+        # For GRUB 0.97 compatibility
+        mkdir -p /mnt/boot/grub
+        echo '(hd0) /dev/vda' > /mnt/boot/grub/device.map
```
Not sure what the plans are to yet out grub without 2.

RaitoBezarius replied:
I think we are safe to remove it, but wanted to get some opinions on this. No one complained for some time now, I guess this can be removed and we can revert if it breaks boot for someone.

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+There are two operations mode for this function:
+
+- either, you want only a Nix store disk image, it can be done using `cptofs` without any virtual machine involved ;
+- either, you want a full NixOS install on the disk, it will start a virtual machine to perform various tasks such as partitionning, installation and bootloader installation.
```
```suggestion
- or, you want a full NixOS install on the disk, it will start a virtual machine to perform various tasks such as partitionning, installation and bootloader installation.
```

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+- either, you want only a Nix store disk image, it can be done using `cptofs` without any virtual machine involved ;
+- either, you want a full NixOS install on the disk, it will start a virtual machine to perform various tasks such as partitionning, installation and bootloader installation.
+
+When NixOS tests are running, this function gets called to generate potentially two kinds of disk images:
```
Which function?

RaitoBezarius replied:
`make-disk-image`

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+## Features
+
+- whether to produce a Nix-store only image or not, **this can be incompatible with other options**
+- arbitrary NixOS configuration ;
```
```suggestion
- arbitrary NixOS configuration
```
looks weird when lists have semicolon

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+
+- bootloader installation have timestamps ;
+- SQLite Nix store database contain registration times ;
+- UIDs / GIDs Nix mappings are in a non-deterministic order ;
```
```suggestion
```
that should be fixed soon by a PR of mine

RaitoBezarius replied:
I think this is fixed now indeed

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+
+- whether to produce a Nix-store only image or not, **this can be incompatible with other options**
+- arbitrary NixOS configuration ;
+- multiple partition table layouts: EFI, legacy, legacy + GPT, hybrid, none ;
```
We should mention the variable how to change that.

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+
+It relies on the [LKL (Linux Kernel Library) project](https://github.com/lkl/linux) which provides Linux kernel as userspace library.
+
+Note that Nix-store only image only need to run LKL tools to produce an image and will never spawn a virtual machine, whereas full images will always require a virtual machine, but also use LKL.
```
```suggestion
::: {.note}
The Nix-store only image only need to run LKL tools to produce an image and will never spawn a virtual machine, whereas full images will always require a virtual machine, but also use LKL.
:::
```

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+Image preparation phase will produce the initial image layout in a folder:
+
+- devise a root folder based on `$PWD`
+- preparing the contents by copying and restoring ACLs in this root folder ;
+- load in the Nix store database all additional paths computed by `pkgs.closureInfo` in a temporary Nix store ;
+- run `nixos-install` in a temporary folder ;
+- transfer from the temporary store the additional paths registered to the installed NixOS ;
+- do fancy computations for the size of the disk image based on the apparent size of the root folder ;
+- partition the disk image using the corresponding script according to the partition table type ;
+- format the partitions if needed ;
```
```suggestion
- preparing the contents by copying and restoring ACLs in this root folder
- load in the Nix store database all additional paths computed by `pkgs.closureInfo` in a temporary Nix store
- run `nixos-install` in a temporary folder
- transfer from the temporary store the additional paths registered to the installed NixOS
- do fancy computations for the size of the disk image based on the apparent size of the root folder
- partition the disk image using the corresponding script according to the partition table type
- format the partitions if needed
```

SuperSandro2000 commented on doc/builders/images/makediskimage.section.md:
```diff
@@ -0,0 +1,180 @@
+This partition table type uses GPT and:
+
+- create a "no filesystem" partition from 1MiB to 2MiB ;
+- set `bios_grub` flag on this "no filesystem" partition, which is some sort of MBR ;
```
Can we clarify what this actually is?

roberth commented:
Please have a look at the new build failures in https://hydra.nixos.org/eval/1783711#tabs-now-fail

RaitoBezarius commented:
Thanks a lot, it seems there are some jobs that are failing on master and
are not related to the PR, do you know how reliable is Hydra about this?

Le mar. 18 oct. 2022 à 20:40, Robert Hensing ***@***.***> a
écrit :

> Please have a look at the new build failures in
> https://hydra.nixos.org/eval/1783711#tabs-now-fail
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/191665#issuecomment-1282850664>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACMZRGFDSF5M77MWAYJ5ILWD3VITANCNFSM6AAAAAAQPAWCTU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

roberth commented:
In my experience, a difference of about a dozen in either direction can be expected because of flaky tests. (when iterating on just the tests; not mass rebuilds.)

This one is definitely disk related: https://hydra.nixos.org/build/195762585/nixlog/12
This is a No such file or directory error, so that could also be image related: https://hydra.nixos.org/build/195762446/nixlog/3
This has a firmware error; could be EFI related: https://hydra.nixos.org/build/195764414/nixlog/1

RaitoBezarius commented:
@sikmir why did you merge the PR?

sikmir commented:
Sorry, I'm trying to figure out why it merged, while I merged anoher PR. I really wonder.

RaitoBezarius commented:
> Sorry, I'm trying to figure out why it merged, while I merged anoher PR. I really wonder.

Ha, okay, I see what is the revision now.

sikmir commented:
But I don't see merge commit on master. What's going on?

RaitoBezarius commented:
> But I don't see merge commit on master. What's going on? Why this PR have no changes?

I suppose this is a GitHub bug due to the fact this PR was "empty" at some point and GitHub misunderstood.

sikmir commented:
> > But I don't see merge commit on master. What's going on? Why this PR have no changes?
> 
> I suppose this is a GitHub bug due to the fact this PR was "empty" at some point and GitHub misunderstood.

Oh, I hope so. It's really strange.
